execution/aggregate/khashaggregate.go

```diff
@@ -0,0 +1,232 @@
+// Copyright (c) The Thanos Community Authors.
+// Licensed under the Apache License 2.0.
+
+package aggregate
+
+import (
+	"container/heap"
+	"context"
+	"fmt"
+	"math"
+	"sort"
+	"sync"
+
+	"github.com/prometheus/prometheus/model/labels"
+	"github.com/prometheus/prometheus/promql/parser"
+	"golang.org/x/exp/slices"
+
+	"github.com/thanos-community/promql-engine/execution/model"
+)
+
+type kAggregate struct {
+	next    model.VectorOperator
+	paramOp model.VectorOperator
+	// params holds the aggregate parameter for each step.
+	params []float64
+
+	vectorPool *model.VectorPool
+
+	by          bool
+	labels      []string
+	aggregation parser.ItemType
+
+	once        sync.Once
+	series      []labels.Labels
+	inputToHeap []*samplesHeap
+	heaps       []*samplesHeap
+	compare     func(float64, float64) bool
+}
+
+func NewKHashAggregate(
+	points *model.VectorPool,
+	next model.VectorOperator,
+	paramOp model.VectorOperator,
+	aggregation parser.ItemType,
+	by bool,
+	labels []string,
+	stepsBatch int,
+) (model.VectorOperator, error) {
+	var compare func(float64, float64) bool
+
+	if aggregation == parser.TOPK {
+		compare = func(f float64, s float64) bool {
+			return f < s
+		}
+	} else {
+		compare = func(f float64, s float64) bool {
+			return s < f
+		}
+	}
+	// Grouping labels need to be sorted in order for metric hashing to work.
+	// https://github.com/prometheus/prometheus/blob/8ed39fdab1ead382a354e45ded999eb3610f8d5f/model/labels/labels.go#L162-L181
+	slices.Sort(labels)
+
+	a := &kAggregate{
+		next:        next,
+		vectorPool:  points,
+		by:          by,
+		aggregation: aggregation,
+		labels:      labels,
+		paramOp:     paramOp,
+		compare:     compare,
+		params:      make([]float64, stepsBatch),
+	}
+
+	return a, nil
+}
+
+func (a *kAggregate) Next(ctx context.Context) ([]model.StepVector, error) {
+	in, err := a.next.Next(ctx)
+	if err != nil {
+		return nil, err
+	}
+	if in == nil {
+		return nil, nil
+	}
+
+	defer a.next.GetPool().PutVectors(in)
+
+	args, err := a.paramOp.Next(ctx)
+	if err != nil {
+		return nil, err
+	}
+	for i := range a.params {
+		a.params[i] = math.NaN()
+		if i < len(args) {
+			a.params[i] = args[i].Samples[0]
+			a.paramOp.GetPool().PutStepVector(args[i])
+		}
+	}
+	a.paramOp.GetPool().PutVectors(args)
+
+	if len(args) != len(in) {
```

Comment:
I think we can remove this now because we inject `NaN`s when a parameter value is not found.